### DISCUSSION TITLE
Delaying & Unique queueable action

### DISCUSSION THREAD

**Discussion by gregory-claeyssens** [spatie/laravel-queueable-action#90] [Ideas]
Would it make sense to add the ability to delay a queued action and ensuring it's unique.
Laravel Queues offer this functionality out of the box (see [delays](https://laravel.com/docs/9.x/queues#delayed-dispatching) and [unique](https://laravel.com/docs/9.x/queues#unique-jobs)).

**Comment by jacksontong** [comment]
You can wrap your action in `ActionJob` to delay the execution

```php
ActionJob::dispatch(MyAction::class, $args)
    ->delay(now()->addSeconds(10));
```

**Reply by bubunamawu** [reply]
Another approach

```
class ActionClass
{
    use QueueableAction;

    public \DateTimeInterface|\DateInterval|int|null  $delay = null;

    public function execute($args) : void
    {
        //do something
    }

    public function delay(\DateTimeInterface|\DateInterval|int|null  $delay) : self
    {
        $this->delay = $delay;
        return $this;
    }
}
```

Then else where
```
resolve(ActionClass::class)->delay(now()->addMinute(2))->onQueue()->execute($args)
```

**Comment by wsamoht** [comment]
Did you ever figure out a way to use [Unique Jobs](https://laravel.com/docs/12.x/queues#unique-jobs) with actions?

EDIT: Looks like a PR was attempted - https://github.com/spatie/laravel-queueable-action/pull/95
